src/coprocessor/codec/convert.rs

```diff
@@ -243,23 +404,53 @@ fn get_valid_float_prefix<'a>(ctx: &mut EvalContext, s: &'a str) -> Result<&'a s
             valid_len = i + 1;
         }
     }
-    box_try!(ctx.handle_truncate(valid_len < s.len()));
+    ctx.handle_truncate(valid_len < s.len())?;
     if valid_len == 0 {
         Ok("0")
     } else {
         Ok(&s[..valid_len])
     }
 }
 
+fn round_int_str(num_next_dot: char, s: &str) -> Cow<'_, str> {
+    if num_next_dot < '5' {
+        return Cow::Borrowed(s);
+    }
+
+    let mut int_str = String::with_capacity(s.len());
+    match s.rfind(|c| c != '9' && c != '+' && c != '-') {
+        Some(idx) => {
+            int_str.push_str(&s[..idx]);
+            let next_char = char::from_u32(s.chars().nth(idx).unwrap() as u32 + 1).unwrap();
+            int_str.push(next_char);
+            let zero_count = s.len() - (idx + 1);
+            if zero_count > 0 {
+                int_str.extend((0..zero_count).map(|_| '0'));
+            }
+        }
+        None => {
+            let zero_count = if s.starts_with('+') || s.starts_with('-') {
+                int_str.push_str(&s[..1]);
+                s.len() - 1
+            } else {
+                s.len()
+            };
+            int_str.push('1');
+            int_str.extend((0..zero_count).map(|_| '0'));
+        }
+    }
+    Cow::Owned(int_str)
+}
+
 /// It converts a valid float string into valid integer string which can be
 /// parsed by `i64::from_str`, we can't parse float first then convert it to string
 /// because precision will be lost.
 ///
 /// When the float string indicating a value that is overflowing the i64,
 /// the original float string is returned and an overflow warning is attached
-fn float_str_to_int_string<'a, 'b: 'a>(
+fn float_str_to_int_string<'a>(
     ctx: &mut EvalContext,
-    valid_float: &'b str,
+    valid_float: &'a str,
 ) -> Result<Cow<'a, str>> {
     let mut dot_idx = None;
     let mut e_idx = None;
```

Comment:
Looks like there is no such logic in TiDB:

```
            '0'..='9' => {
                if e_idx.is_none() {
                    if dot_idx.is_none() {
                        int_cnt += 1;
                    }
                    digits_cnt += 1;
                }
            }
```

Reply:
The original use different way to implement the logic. I kept it and fixed some corner case.

Reply:
For the first version I prefer to keep the code identical as much as possible, especially considering that we have proposed many changes in this PR. I don't think we have tests that can cover 100% branches so that differences are likely to cause many kind of bugs.